src/bin/nydus-image/core/node.rs

```diff
@@ -1150,31 +1152,61 @@ impl Node {
         //
         //
         let inode_size = self.size_with_xattr() as u64;
-
-        bootstrap_ctx.offset += inode_size;
-        let avail: u64 = EROFS_BLOCK_SIZE - bootstrap_ctx.offset % EROFS_BLOCK_SIZE;
         let tail: u64 = d_size % EROFS_BLOCK_SIZE;
-        self.v6_datalayout = if tail > 0 {
-            if avail >= tail && bootstrap_ctx.offset % EROFS_BLOCK_SIZE as u64 > 0 {
-                bootstrap_ctx.offset += tail;
-                if d_size > tail {
-                    // add remained bytes of 'd_size' in a new 4k.
+
+        // We use a simple inline strategy here:
+        // If the inode size with xattr + tail data size <= EROFS_BLOCK_SIZE,
+        // we choose to inline it.
+        // Firstly, if it's bigger that EROFS_BLOCK_SIZE,
+        // in most cases, we can assume that the tail data size is close to EROFS_BLOCK_SIZE,
+        // in this condition, even if we don't inline the tail data, there won't be much waste.
+        // Secondly, the `available_blocks` that we maintain in the `BootstrapCtx`,
+        // since it contain only single blocks with some unsed space, the available space can only be smaller that EROFS_BLOCK_SIZE,
+        // therefore we can't use our used blocks to store the inode plus the tail data bigger than EROFS_BLOCK_SIZE.
+        let should_inline = tail != 0 && (inode_size + tail) <= EROFS_BLOCK_SIZE;
```

Comment:
how about inode_size + inline xattr + tail > EROFS_BLOCK_SIZE but inline xattr % 4k + tail < EROFS_BLOCK_SIZE?

Reply:
I suggest that you need to add some test cases for these, also you need to consider if nydus will use EROFS compressed inodes as well, so generic logic is preferred.

Reply:
`compressed inodes`?
does it refer to `compact inodes`?

@hsiangkao

Reply:
no, I was thinking we would add compressed inodes maybe later, and use flat inodes not only for directory inodes.
The current approach is not quite flexible on my side.